src/service/link-rewriter/link-rewriter.js

```diff
@@ -0,0 +1,239 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {user} from '../../log';
+
+import {EVENTS, ORIGINAL_URL_ATTRIBUTE} from './constants';
+import {EventMessenger} from './event-messenger';
+import {LinkReplacementCache} from './link-replacement-cache';
+import {isTwoStepsResponse} from './link-rewriter-helpers';
+
+
+/** @typedef {!Array<{anchor: HTMLElement, replacementUrl: ?string}>}} */
+export let AnchorReplacementList;
+
+/** @typedef {{syncResponse: ?AnchorReplacementList, asyncResponse: ?Promise<!AnchorReplacementList>}} */
+export let TwoStepsResponse;
+
+
+/**
+ * LinkRewriter works together with LinkRewriterManager to allow rewriting
+ * links at click time. E.g: Replacing a link by its affiliate version only if
+ * the link can be monetised.
+ *
+ * A page can have multiple LinkRewriter running at the same time.
+ *
+ * LinkRewriter class is in charge of:
+ * - Scanning the page to find links based on an optional CSS selector.
+ *
+ * - Asking the "resolveUnknownLinks" function which links can be replaced
+ *   so we already know the replacement URL at click time.
+ *
+ * - Keeping track of the replacement link of each links on the page.
+ *
+ * - Swapping the anchor url to its replacement url when instructed
+ *   by the LinkRewriterManager.
+ */
+export class LinkRewriter {
```

Comment:
it looks to me this service serves very similar purpose as the existing navigation.js. shall we consolidate? /cc @alabiaga

Reply:
Yes, this logic should just be pushed to the extension itself so as not to bloat the runtime binary. The service can then import and use the navigation service to register the link rewrite logic via registerAnchorMutator. This should be done to respect the URL variable expansion logic that occurs in the navigation service and to ensure that it occurs prior to when the links are mutated. Each registered link rewriter/anchor mutator will then have a reference to what the URL would be originally. The url expansion perhaps is not relevant in determining whether a link is monetizable but the whole purpose of adding this mutator logic in navigation.js is so that we have a central place of handling anchor mutations that occur in AMP runtime, as well as ensuring an order to which these mutations occur. Please contact @calebcordry as to how to use this, as he has used it recently in his extension.

Note that I noticed there has been a lot of thought put into your link rewriter and manager interactions and we took a simpler approach in the navigation service; each extension registering a link rewriter/anchor mutator is required to reserve a priority which will be respected. I was under the assumption that we all agreed that the amp skim links extension was going to do a simple click event listener to do its required logic, as we developed the logic to support this in runtime? We can keep most of this logic in the extension and port it to the AMP runtime if the need for it is necessary. Thanks

Reply:
The `LinkRewriter` solution was an attempt for us to contribute to the amp project beyond amp-skimlinks extension by generalising the link rewriting part of affiliate services. Anchor mutators didn't exists in `navigation.js` at the time this code was written and we decided to take a similar approach to fix potential conflicts between click handlers trying to mutate the anchor.

I will move the LinkRewriter logic back to amp-skimlinks code and will plug it to the new anchorMutator mechanism. We could always try to generalise it later.

Reply:
Sounds good. Many thanks Vincent.